### ISSUE TITLE
libssh2 incorrectly uses openssl 1.0.x symbols when built against openssl 1.1.0

### ISSUE THREAD

**Issue by pgnd** [libssh2/libssh2#175]
I've doc'd the problem here

**libcurl HEAD build w/ Openssl 1.1 API fails at incorrect check for Openssl 1.0 API symbols in libssh2 #1309**
https://github.com/curl/curl/issues/1309#issuecomment-284270734

and applied a crude patch that works around the issue:

```
diff -ur /usr/local/src/libssh2.ORIG/src/openssl.c /usr/local/src/libssh2/src/openssl.c
--- /usr/local/src/libssh2.ORIG/src/openssl.c   2017-03-04 12:37:07.225631493 -0800
+++ /usr/local/src/libssh2/src/openssl.c        2017-03-05 14:40:41.039767506 -0800
@@ -1062,7 +1062,6 @@
          * ($#&%#$(%$#( Someone buy me an OpenSSL manual and I'll read up on
          * it.
          */
-        OpenSSL_add_all_ciphers();
     }
     BIO_reset(bp);
     pk = PEM_read_bio_PrivateKey(bp, NULL, NULL, (void*)passphrase);
@@ -1138,7 +1137,6 @@
          * ($#&%#$(%$#( Someone buy me an OpenSSL manual and I'll read up on
          * it.
          */
-        OpenSSL_add_all_ciphers();
     }
     BIO_reset(bp);
     pk = PEM_read_bio_PrivateKey(bp, NULL, NULL, (void*)passphrase);
diff -ur /usr/local/src/libssh2.ORIG/src/openssl.h /usr/local/src/libssh2/src/openssl.h
--- /usr/local/src/libssh2.ORIG/src/openssl.h   2017-03-05 14:40:01.343938851 -0800
+++ /usr/local/src/libssh2/src/openssl.h        2017-03-05 14:41:05.891660235 -0800
@@ -227,7 +227,6 @@
 #endif

 #define libssh2_crypto_init() \
-  OpenSSL_add_all_algorithms(); \
   ENGINE_load_builtin_engines(); \
   ENGINE_register_all_complete()
```

clearly, not a final solution.

**Comment by jay** [comment]
That solution will cause problems. You should document the main problem here for them instead of referring to the other thread. My understanding is basically you have compiled libssh2 that is for some reason attempting to find these symbols even though you've built against OpenSSL 1.1
~~~
	/usr/local/lib64/libssh2.so: undefined reference to `OpenSSL_add_all_ciphers'
	/usr/local/lib64/libssh2.so: undefined reference to `OpenSSL_add_all_algorithms'
~~~
So the question is why that is, because in 1.1 OpenSSL_add_all_ciphers is defined as [this](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/include/openssl/evp.h#L861-L862)
~~~
#  define OpenSSL_add_all_ciphers() \
    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS, NULL)
~~~
similar case for the other one. So if the includes were really OpenSSL 1.1 it would be looking for OPENSSL_init_crypto. Therefore I think an older set of includes were used.

**Comment by pgnd** [comment]
> You should document the main problem here for them instead of referring to the other thread

```
cd src/libssh2
git clean -xfd
git reset --hard

git log | head
	commit 323aa08f01489dd3ab0aa11eca16f58ab13a8ff0
	Author: Marc Hoersken <info@marc-hoersken.de>
	Date:   Thu Mar 2 22:29:14 2017 +0100

	    tests/openssh_server/authorized_keys: add key_rsa_encrypted.pub

	commit 20a545d37d0da1148a0582a3789bdb374ed51f80
	Author: Marc Hoersken <info@marc-hoersken.de>
	Date:   Thu Mar 2 00:34:06 2017 +0100

rm -rf build
mkdir  build
cd     build

cmake .. \
 -DCMAKE_INSTALL_PREFIX=/usr/local \
 -DCMAKE_INSTALL_LIBDIR=lib64 \
 -DBUILD_SHARED_LIBS=1 \
 -DCMAKE_BUILD_TYPE=Release \
   -DCMAKE_SKIP_INSTALL_RPATH=0 \
   -DCMAKE_SKIP_RPATH=0 \
   -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=1 \
 -DCRYPTO_BACKEND=OpenSSL \
 -DCMAKE_EXE_LINKER_FLAGS="-L/usr/local/openssl11/lib64 -Wl,-rpath,/usr/local/openssl11/lib64 -lssl -lcrypto -L/usr/local/lib64 -Wl,-rpath,/usr/local/lib64 -lssh2" \
 -DCMAKE_SHARED_LINKER_FLAGS="-L/usr/local/openssl11/lib64 -Wl,-rpath,/usr/local/openssl11/lib64 -lssl -lcrypto -L/usr/local/lib64 -Wl,-rpath,/usr/local/lib64 -lssh2" \
 -DOPENSSL_INCLUDE_DIR=/usr/local/openssl11/include \
 -DOPENSSL_SSL_LIBRARY=/usr/local/openssl11/lib64/libssl.so \
 -DOPENSSL_CRYPTO_LIBRARY=/usr/local/openssl11/lib64/libcrypto.so \
 -DENABLE_CRYPT_NONE=0 \
 -DBUILD_EXAMPLES=0 \
 -DBUILD_TESTING=0

	...
	-- Found OpenSSL: /usr/local/openssl11/lib64/libssl.so;/usr/local/openssl11/lib64/libcrypto.so (found version "1.1.0e") 
	...
	-- The following REQUIRED packages have been found:

	 * OpenSSL

	...
	-- Configuring done
	-- Generating done
	-- Build files have been written to: /usr/local/src/libssh2/build

make V=1
	Scanning dependencies of target libssh2
	[  4%] Building C object src/CMakeFiles/libssh2.dir/comp.c.o
	[  9%] Building C object src/CMakeFiles/libssh2.dir/agent.c.o
	[ 13%] Building C object src/CMakeFiles/libssh2.dir/crypt.c.o
	[ 18%] Building C object src/CMakeFiles/libssh2.dir/channel.c.o
	[ 22%] Building C object src/CMakeFiles/libssh2.dir/openssl.c.o
	[ 27%] Building C object src/CMakeFiles/libssh2.dir/global.c.o
	/usr/local/src/libssh2/src/openssl.c: In function ‘_libssh2_pub_priv_keyfile’:
	/usr/local/src/libssh2/src/openssl.c:1065:9: warning: implicit declaration of function ‘OpenSSL_add_all_ciphers’ [-Wimplicit-function-declaration]
	         OpenSSL_add_all_ciphers();
	         ^~~~~~~~~~~~~~~~~~~~~~~
	In file included from /usr/local/src/libssh2/src/crypto.h:42:0,
	                 from /usr/local/src/libssh2/src/libssh2_priv.h:140,
	                 from /usr/local/src/libssh2/src/global.c:39:
	/usr/local/src/libssh2/src/global.c: In function ‘libssh2_init’:
	/usr/local/src/libssh2/src/openssl.h:230:3: warning: implicit declaration of function ‘OpenSSL_add_all_algorithms’ [-Wimplicit-function-declaration]
	   OpenSSL_add_all_algorithms(); \
	   ^
	/usr/local/src/libssh2/src/global.c:48:9: note: in expansion of macro ‘libssh2_crypto_init’
	         libssh2_crypto_init();
	         ^~~~~~~~~~~~~~~~~~~
	[ 31%] Building C object src/CMakeFiles/libssh2.dir/hostkey.c.o
	[ 36%] Building C object src/CMakeFiles/libssh2.dir/keepalive.c.o
	[ 40%] Building C object src/CMakeFiles/libssh2.dir/kex.c.o
	[ 45%] Building C object src/CMakeFiles/libssh2.dir/knownhost.c.o
	[ 50%] Building C object src/CMakeFiles/libssh2.dir/mac.c.o
	[ 54%] Building C object src/CMakeFiles/libssh2.dir/misc.c.o
	[ 59%] Building C object src/CMakeFiles/libssh2.dir/packet.c.o
	[ 63%] Building C object src/CMakeFiles/libssh2.dir/pem.c.o
	[ 68%] Building C object src/CMakeFiles/libssh2.dir/publickey.c.o
	[ 72%] Building C object src/CMakeFiles/libssh2.dir/scp.c.o
	[ 77%] Building C object src/CMakeFiles/libssh2.dir/session.c.o
	[ 81%] Building C object src/CMakeFiles/libssh2.dir/sftp.c.o
	[ 86%] Building C object src/CMakeFiles/libssh2.dir/transport.c.o
	[ 90%] Building C object src/CMakeFiles/libssh2.dir/userauth.c.o
	[ 95%] Building C object src/CMakeFiles/libssh2.dir/version.c.o
	[100%] Linking C shared library libssh2.so
	[100%] Built target libssh2

make install

ldd /usr/local/lib64/libssh2.so | egrep "ssl|crypto"
	libssl.so.1.1 => /usr/local/openssl11/lib64/libssl.so.1.1 (0x00007fab5123a000)
	libcrypto.so.1.1 => /usr/local/openssl11/lib64/libcrypto.so.1.1 (0x00007fab50d90000)

nm /usr/local/lib64/libssh2.so | grep OpenSSL_add_all
                 U OpenSSL_add_all_algorithms
                 U OpenSSL_add_all_ciphers

make clean

( PATCH (as above) ==> /usr/local/src/libssh2-openssl11.patch ) 

cd ../
patch -p5 < /usr/local/src/libssh2-openssl11.patch
	patching file src/openssl.c
	patching file src/openssl.h
cd build/

make V=1
	[  4%] Building C object src/CMakeFiles/libssh2.dir/openssl.c.o
	[  9%] Building C object src/CMakeFiles/libssh2.dir/agent.c.o
	[ 13%] Building C object src/CMakeFiles/libssh2.dir/channel.c.o
	[ 18%] Building C object src/CMakeFiles/libssh2.dir/comp.c.o
	[ 22%] Building C object src/CMakeFiles/libssh2.dir/crypt.c.o
	[ 27%] Building C object src/CMakeFiles/libssh2.dir/global.c.o
	[ 31%] Building C object src/CMakeFiles/libssh2.dir/hostkey.c.o
	[ 36%] Building C object src/CMakeFiles/libssh2.dir/keepalive.c.o
	[ 40%] Building C object src/CMakeFiles/libssh2.dir/kex.c.o
	[ 45%] Building C object src/CMakeFiles/libssh2.dir/knownhost.c.o
	[ 50%] Building C object src/CMakeFiles/libssh2.dir/mac.c.o
	[ 54%] Building C object src/CMakeFiles/libssh2.dir/misc.c.o
	[ 59%] Building C object src/CMakeFiles/libssh2.dir/packet.c.o
	[ 63%] Building C object src/CMakeFiles/libssh2.dir/pem.c.o
	[ 68%] Building C object src/CMakeFiles/libssh2.dir/publickey.c.o
	[ 72%] Building C object src/CMakeFiles/libssh2.dir/scp.c.o
	[ 77%] Building C object src/CMakeFiles/libssh2.dir/session.c.o
	[ 81%] Building C object src/CMakeFiles/libssh2.dir/sftp.c.o
	[ 86%] Building C object src/CMakeFiles/libssh2.dir/transport.c.o
	[ 90%] Building C object src/CMakeFiles/libssh2.dir/userauth.c.o
	[ 95%] Building C object src/CMakeFiles/libssh2.dir/version.c.o
	[100%] Linking C shared library libssh2.so
	[100%] Built target libssh2

make install

ldd /usr/local/lib64/libssh2.so | egrep "ssl|crypto"
	libssl.so.1.1 => /usr/local/openssl11/lib64/libssl.so.1.1 (0x00007fab5123a000)
	libcrypto.so.1.1 => /usr/local/openssl11/lib64/libcrypto.so.1.1 (0x00007fab50d90000)
nm /usr/local/lib64/libssh2.so | grep OpenSSL_add_all
	(empty)
```

**Comment by bagder** [comment]
As a comparison, I built with `configure`:

    $ nm src/.libs/libssh2.so.1.0.1 | egrep '(OPENSSL_init_crypto|OpenSSL_add_all_)'
           U OPENSSL_init_crypto@@OPENSSL_1_1_0

**Comment by pgnd** [comment]
here,
```
nm /usr/local/lib64/libssh2.so.1.0.1  | egrep "OPENSSL_init_crypto|OpenSSL_add_all"
  (empty)
```
your build was, iiuc, against openssl 1.1.1,

> These were all git master as of now. OpenSSL reports itself as 1.1.1 there.

rt?

**Comment by bagder** [comment]
> your build was, iiuc, against openssl 1.1.1

openssl git master, yes, but as @jay already showed, the #define was there in the openssl 1.1.0 headers already so git master or 1.1.0 shouldn't matter in that regard.

> rt?

Sorry I don't know what that means.

**Comment by MarcelRaad** [comment]
Sorry I had missed this issue when opening PR #176 , which fixes the same problem. 

@jay Those function-style macros only exist if you build OpenSSL with api=1.0 or lower, which sets OPENSSL_API_COMPAT accordingly in opensslconf.h.

**Comment by pgnd** [comment]
Confirming cherry pick of @MarcelRaad 's commit, applied to current head,
```
	git log | head -n1
		commit 323aa08f01489dd3ab0aa11eca16f58ab13a8ff0
	git remote add MarcelRaad https://github.com/MarcelRaad/libssh2.git
	git fetch MarcelRaad
	git cherry-pick 61c1158dc5627c473dbe0edae6ff0210710f0bfd
```
builds correctly linked libssh2 lib
```
	ldd /usr/local/lib64/libssh2.so | egrep "ssl|crypto"
		libssl.so.1.1 => /usr/local/openssl11/lib64/libssl.so.1.1 (0x00007f53c3015000)
		libcrypto.so.1.1 => /usr/local/openssl11/lib64/libcrypto.so.1.1 (0x00007f53c2b6b000)
```
and makes ok in downstream libcurl, 
```
	ldd ./lib/.libs/libcurl.so | egrep "libssl|libcrypto|libssh|mysql"
		libssl.so.1.1 => /usr/local/openssl11/lib64/libssl.so.1.1 (0x00007f2e93a90000)
		libcrypto.so.1.1 => /usr/local/openssl11/lib64/libcrypto.so.1.1 (0x00007f2e935e6000)
		libssh2.so.1 => /usr/local/lib64/libssh2.so.1 (0x00007f2e933b3000)

	which curl
		/usr/local/bin/curl
	curl -V
		curl 7.54.0-DEV (x86_64-suse-linux-gnu) libcurl/7.54.0-DEV OpenSSL/1.1.0e zlib/1.2.11 libpsl/0.17.0 (+libicu/52.1) libssh2/1.8.1_DEV nghttp2/1.21.0-DEV
		Protocols: dict file ftp ftps http https imap imaps pop3 pop3s scp sftp smtp smtps tftp
		Features: AsynchDNS IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy Metalink PSL

```
which was the original point of this^ ...

**Comment by pgnd** [comment]
as it's been a couple of months, trying again with latest head
```
	cd src/libssh2
	git log | head
		commit a9179fb25709983dea727e54b005add1ad55464e
		Merge: 94fd8ed 97518ca
		Author: root <root@deskEdgar.presence-group.net>
		Date:   Mon May 22 10:24:12 2017 -0700

		    Merge branch 'master' of https://github.com/libssh2/libssh2

		commit 97518ca8bda91ce12c503197a98fa71690cb67f9
		Author: Will Cosgrove <will@panic.com>
		Date:   Fri May 12 16:34:26 2017 -0700
```
building against OpenSSL 1.1 api libs
```
	cmake \
	 ...
	 -DCRYPTO_BACKEND=OpenSSL \
	 -DCMAKE_EXE_LINKER_FLAGS="-L/usr/local/openssl11/lib64 -Wl,-rpath,/usr/local/openssl11/lib64 -lssl -lcrypto -L/usr/local/lib64 -Wl,-rpath,/usr/local/lib64 -lssh2" \
	 -DCMAKE_SHARED_LINKER_FLAGS="-L/usr/local/openssl11/lib64 -Wl,-rpath,/usr/local/openssl11/lib64 -lssl -lcrypto -L/usr/local/lib64 -Wl,-rpath,/usr/local/lib64 -lssh2" \
	 -DOPENSSL_INCLUDE_DIR=/usr/local/openssl11/include \
	 -DOPENSSL_SSL_LIBRARY=/usr/local/openssl11/lib64/libssl.so \
	 -DOPENSSL_CRYPTO_LIBRARY=/usr/local/openssl11/lib64/libcrypto.so \
	 ...
	make V=1
```
configs & builds OK
```
	...
	[100%] Linking C shared library libssh2.so
	[100%] Built target libssh2
```
but still leaves lib symbols undef'd
```
	nm /usr/local/lib64/libssh2.so | grep OpenSSL_add_all
		U OpenSSL_add_all_algorithms
		U OpenSSL_add_all_ciphers
```
Cherry-pick from @MarcelRaad's branch
```
	git remote add MarcelRaad https://github.com/MarcelRaad/libssh2.git
	git fetch MarcelRaad
	git cherry-pick 61c1158dc5627c473dbe0edae6ff0210710f0bfd
	git log | head
		commit 988df71f41f2534a6613291b3467c860fccf0980
		Author: Marcel Raad <raad@teamviewer.com>
		Date:   Wed Mar 8 11:46:44 2017 +0100

		    openssl: fix build with OpenSSL 1.1 API

		    When building with OPENSSL_API_COMPAT=0x10100000L, OpenSSL_add_all_algorithms
		    and OpenSSL_add_all_ciphers don't exist. The corresponding functionality is
		    handled automatically with OpenSSL 1.1.
```
still cures the problem, with NO undef'd symbols in built/installed libs
```
	... as above ...
	nm /usr/local/lib64/libssh2.so | grep OpenSSL_add_all
		(empty)
```
What further info is required to get this promoted into master?

**Comment by willco007** [comment]
Looked into this; as @MarcelRaad notes, the OpenSSL macros are only defined if you build with API compatibility enabled which we can't guarantee (thus the build issue).  The PR seems safe enough; any objections to applying it?

**Comment by willco007** [comment]
Merged PR #176 which resolves this issue.
